File: py-proio/proio/event.py
import proio.proto as proto

import google.protobuf.descriptor_pool as descriptor_pool
import google.protobuf.message_factory as message_factory

class Event(object):
    """
    Class representing a single event
    """

    def __init__(self, proto_obj = None):
        self._proto = proto_obj or proto.Event()
        self._entry_cache = {}
        self._factory = message_factory.MessageFactory()
        self._rev_type_lookup = {}
        self._dirty_tags = False

    def add_entry(self, tag, entry):
        """
        takes a tag and protobuf message entry and adds it to the Event.  The
        return value is an integer ID number used to reference the added entry.

        :param string tag: tag
        :param Message entry: entry
        :return: identifier for entry
        :rtype: int
        """
        type_id = self._get_type_id(entry)

        self._proto.nEntries += 1
        ID = self._proto.nEntries
        self._proto.entries[ID].type = type_id

        self._entry_cache[ID] = entry

        self.tag_entry(ID, tag)

        return ID

    def add_entries(self, tag, *entries):
        """
        is like :func:`add_entry`, except that it takes any number of entries,
        and returns a list of corresponding IDs.

        :param string tag: tag
        :param Message \*entries: entries
        :return: identifiers for entry
        :rtype: int list
        """
        ids = []
        for entry in entries:
            ids.append(self.add_entry(tag, entry))
        return ids

    def get_entry(self, ID):
        """
        takes an entry ID and returns the corresponding entry.  Returns None if
        the entry does not exist.

        :param int ID: identifier for entry
        :return: entry message object
        :rtype: :class:`google.protobuf.message.Message`
        """
        if ID in self._entry_cache:
            return self._entry_cache[ID]
        if ID in self._proto.entries:
            entry_proto = self._proto.entries[ID]
        else:
            return None

        type_string = self._proto.types[entry_proto.type]
        type_desc = descriptor_pool.Default().FindMessageTypeByName(type_string)
        msg_class = self._factory.GetPrototype(type_desc)
        entry = msg_class.FromString(entry_proto.payload)
        self._entry_cache[ID] = entry

        return entry

    def remove_entry(self, ID):
        """
        removes an entry from the event by its entry ID.

        :param int ID: identifier for entry
        """
        self._entry_cache.pop(ID, None)
        self._proto.entries.pop(ID, None)
        self._dirty_tags = True

    def tag_entry(self, ID, *tags):
        """
        adds tags to an entry identified by ID.

        :param int ID: identifier for entry
        :param string \*tags: tags
        """
        for tag in tags:
            tag_proto = self._proto.tags[tag]
            tag_proto.entries.append(ID)

    def untag_entry(self, ID, tag):
        """
        removes the specified tag from the entry ID

        :param int ID: identifier for entry
        :param string tag: tag
        """
        if tag in self._proto.tags:
            try:
                self._proto.tags[tag].entries.remove(ID)
            except ValueError:
                pass

    def tags(self):
        """
        returns a list of tags that exist in the event.

        :return: list of strings
        """
        tags = list(self._proto.tags.keys())
        tags.sort()
        return tags

    def entry_tags(self, ID):
        """
        returns a list of tags that point to a given entry ID.

        :param int ID: identifier for entry
        :return: list of strings
        """
        tags = []
        for tag, value in self._proto.tags.items():
            if ID in value.entries:
                tags.append(tag)
        return tags

    def tagged_entries(self, tag):
        """
        takes a tag string and returns a list of entry IDs that the tag
        references.

        :param string tag: tag
        :return: identifiers for entries
        :rtype: int list
        """
        self._tag_cleanup()
        return list(self._proto.tags[tag].entries)

    def all_entries(self):
        """
        returns a list of entry IDs for the event.

        :return: identifiers for entries
        :rtype: int list
        """
        return list(self._proto.entries.keys())

    def delete_tag(self, tag):
        """
        removes the tag that matches the given tag string from the event.

        :param string tag: tag
        """
        self._proto.tags.pop(tag, None)

    def _get_type_id(self, entry):
        type_name = entry.DESCRIPTOR.full_name
        try:
            return self._rev_type_lookup[type_name]
        except KeyError:
            for ID, name in self._proto.types.items():
                if name == type_name:
                    self._rev_type_lookup[name] = ID
                    return ID

            self._proto.nTypes += 1
            type_id = self._proto.nTypes
            self._proto.types[type_id] = type_name
            self._rev_type_lookup[type_name] = type_id

            return type_id

    def _flush_cache(self):
        for ID, entry in self._entry_cache.items():
            self._proto.entries[ID].payload = entry.SerializeToString()
        self._entry_cache = {}
        self._tag_cleanup()

    def __str__(self):
        print_string = ''

        tags = self.tags()
        for tag in tags:
            print_string += '---------- TAG: ' + tag + ' ----------\n'
            entries = self.tagged_entries(tag)
            for entry_id in entries:
                print_string += 'ID: %i\n' % entry_id
                entry = self.get_entry(entry_id)
                if entry is not None:
                    print_string += 'Entry type: ' + entry.DESCRIPTOR.full_name + '\n'
                    print_string += '%s\n' % entry
                else:
                    print_string += 'not found'

        return print_string

    def _tag_cleanup(self):
        if not self._dirty_tags:
            return
        for tag in self._proto.tags.values():
            tag.entries[:] = [ID for ID in tag.entries if ID in self._proto.entries]
        self._dirty_tags = False

File: go-proio/event.go
package proio // import "github.com/decibelcooper/proio/go-proio"

// Generate protobuf messages
//go:generate bash gen.sh

import (
	"errors"
	"fmt"
	"reflect"
	"sort"
	"strconv"

	"github.com/decibelcooper/proio/go-proio/proto"
	protobuf "github.com/golang/protobuf/proto"
)

// Event contains all data for an event, and provides methods for adding and
// retrieving data.
type Event struct {
	Err      error
	Metadata map[string][]byte

	proto *proto.Event

	revTypeLookup  map[string]uint64
	entryTypeCache map[uint64]reflect.Type
	entryCache     map[uint64]protobuf.Message
	dirtyTags      bool
}

// NewEvent is required for constructing an Event.
func NewEvent() *Event {
	return &Event{
		Metadata: make(map[string][]byte),
		proto: &proto.Event{
			Entry: make(map[uint64]*proto.Any),
			Type:  make(map[uint64]string),
			Tag:   make(map[string]*proto.Tag),
		},
		revTypeLookup:  make(map[string]uint64),
		entryTypeCache: make(map[uint64]reflect.Type),
		entryCache:     make(map[uint64]protobuf.Message),
		dirtyTags:      false,
	}
}

// AddEntry takes a single primary tag for an entry and an entry protobuf
// message, and returns a new ID number for the entry.  This ID number can be
// used to persistently reference the entry.  For example, pass the ID TagEntry
// to add additional tags to the entry.
func (evt *Event) AddEntry(tag string, entry protobuf.Message) uint64 {
	typeID := evt.getTypeID(entry)
	entryProto := &proto.Any{
		Type: typeID,
	}

	evt.proto.NEntries++
	id := evt.proto.NEntries
	evt.proto.Entry[id] = entryProto

	evt.entryCache[id] = entry

	evt.TagEntry(id, tag)

	return id
}

// AddEntries is like AddEntry, except that it is variadic, taking an arbitrary
// number of entries separated by commas.  Additionally, the return value is a
// slice of IDs.
func (evt *Event) AddEntries(tag string, entries ...protobuf.Message) []uint64 {
	ids := make([]uint64, len(entries))
	for i, entry := range entries {
		ids[i] = evt.AddEntry(tag, entry)
	}
	return ids
}

// GetEntry retrieves and deserializes an entry corresponding to the given ID
// number.  The deserialized entry is returned.  The entry type must be one
// that has been linked (and therefore initialized) with the current
// executable, otherwise it is an unknown type and nil is returned.
func (evt *Event) GetEntry(id uint64) protobuf.Message {
	entry, ok := evt.entryCache[uint64(id)]
	if ok {
		evt.Err = nil
		return entry
	}

	entryProto, ok := evt.proto.Entry[uint64(id)]
	if !ok {
		evt.Err = errors.New("no such entry: " + strconv.FormatUint(id, 10))
		return nil
	}

	entry = evt.getPrototype(entryProto.Type)
	if entry == nil {
		evt.Err = errors.New("unknown type: " + evt.proto.Type[entryProto.Type])
		return nil
	}

	if err := protobuf.Unmarshal(entryProto.Payload, entry); err != nil {
		evt.Err = errors.New(
			"failure to unmarshal entry " +
				strconv.FormatUint(id, 10) +
				" with type " +
				evt.proto.Type[entryProto.Type],
		)
		return nil
	}

	evt.entryCache[id] = entry

	evt.Err = nil
	return entry
}

// RemoveEntry takes an entry id and removes the referenced entry from the
// Event.
func (evt *Event) RemoveEntry(id uint64) {
	delete(evt.entryCache, id)
	delete(evt.proto.Entry, id)
	evt.dirtyTags = true
}

// AllEntries returns a slice of identifiers for all entries contained in the
// Event.
func (evt *Event) AllEntries() []uint64 {
	IDs := make([]uint64, len(evt.proto.Entry))
	var i int
	for ID := range evt.proto.Entry {
		IDs[i] = ID
		i++
	}
	return IDs
}

// TagEntry adds additional tags to an entry ID returned by AddEntry.
func (evt *Event) TagEntry(id uint64, tags ...string) {
	for _, tag := range tags {
		tagProto, ok := evt.proto.Tag[tag]
		if !ok {
			tagProto = &proto.Tag{}
			evt.proto.Tag[tag] = tagProto
		}

		tagProto.Entry = append(tagProto.Entry, id)
	}
}

// UntagEntry removes the association between a tag and an entry.
func (evt *Event) UntagEntry(id uint64, tag string) {
	tagProto, ok := evt.proto.Tag[tag]
	if !ok {
		return
	}

	for i, entryID := range tagProto.Entry {
		if entryID == id {
			tagProto.Entry = append(tagProto.Entry[:i], tagProto.Entry[i+1:]...)
			return
		}
	}
}

// TaggedEntries returns a slice of ID numbers that are referenced by the given
// tag.
func (evt *Event) TaggedEntries(tag string) []uint64 {
	tagProto, ok := evt.proto.Tag[tag]
	if ok {
		evt.tagCleanup()
		entries := make([]uint64, len(tagProto.Entry))
		copy(entries, tagProto.Entry)
		return entries
	}
	return nil
}

// Tags returns a list of all tags in the Event.
func (evt *Event) Tags() []string {
	var tags []string
	for key := range evt.proto.Tag {
		tags = append(tags, key)
	}
	sort.Strings(tags)
	return tags
}

// EntryTags does a reverse lookup of tags that point to a given entry ID.
func (evt *Event) EntryTags(id uint64) []string {
	tags := make([]string, 0)
	for name, tagProto := range evt.proto.Tag {
		for _, thisID := range tagProto.Entry {
			if thisID == id {
				tags = append(tags, name)
				break
			}
		}
	}
	sort.Strings(tags)
	return tags
}

// DeleteTag takes a tag name as an argument and deletes that tag from the
// Event if it exists.
func (evt *Event) DeleteTag(tag string) {
	delete(evt.proto.Tag, tag)
}

func (evt *Event) String() string {
	var printString string

	tags := evt.Tags()

	for _, tag := range tags {
		printString += "---------- TAG: " + tag + " ----------\n"
		entries := evt.TaggedEntries(tag)
		for _, entryID := range entries {
			printString += fmt.Sprintf("ID: %v\n", entryID)
			entry := evt.GetEntry(entryID)
			if entry != nil {
				typeName := protobuf.MessageName(entry)
				printString += "Entry type: " + typeName + "\n"
				printString += protobuf.MarshalTextString(entry) + "\n"
			} else {
				printString += evt.Err.Error() + "\n"
			}
		}
	}

	return printString
}

func newEventFromProto(eventProto *proto.Event) *Event {
	if eventProto.Entry == nil {
		eventProto.Entry = make(map[uint64]*proto.Any)
	}
	if eventProto.Type == nil {
		eventProto.Type = make(map[uint64]string)
	}
	if eventProto.Tag == nil {
		eventProto.Tag = make(map[string]*proto.Tag)
	}
	return &Event{
		Metadata:       make(map[string][]byte),
		proto:          eventProto,
		revTypeLookup:  make(map[string]uint64),
		entryTypeCache: make(map[uint64]reflect.Type),
		entryCache:     make(map[uint64]protobuf.Message),
	}
}

func (evt *Event) getPrototype(id uint64) protobuf.Message {
	entryType, ok := evt.entryTypeCache[id]
	if !ok {
		ptrType := protobuf.MessageType(evt.proto.Type[id])
		if ptrType == nil {
			return nil
		}
		entryType = ptrType.Elem()
		evt.entryTypeCache[id] = entryType
	}

	return reflect.New(entryType).Interface().(protobuf.Message)
}

func (evt *Event) getTypeID(entry protobuf.Message) uint64 {
	typeName := protobuf.MessageName(entry)
	typeID, ok := evt.revTypeLookup[typeName]
	if !ok {
		for id, name := range evt.proto.Type {
			if name == typeName {
				evt.revTypeLookup[typeName] = id
				return id
			}
		}

		evt.proto.NTypes++
		typeID = evt.proto.NTypes
		evt.proto.Type[typeID] = typeName
		evt.revTypeLookup[typeName] = typeID
	}

	return typeID
}

func (evt *Event) flushCache() {
	for id, entry := range evt.entryCache {
		selfSerializingEntry, ok := entry.(protobuf.Marshaler)
		var bytes []byte
		if ok {
			bytes, _ = selfSerializingEntry.Marshal()
		} else {
			bytes, _ = protobuf.Marshal(entry)
		}
		evt.proto.Entry[id].Payload = bytes
	}
	evt.entryCache = make(map[uint64]protobuf.Message)

	evt.tagCleanup()
}

func (evt *Event) tagCleanup() {
	if !evt.dirtyTags {
		return
	}
	for _, tagProto := range evt.proto.Tag {
		for i := len(tagProto.Entry) - 1; i >= 0; i-- {
			if _, ok := evt.proto.Entry[tagProto.Entry[i]]; !ok {
				tagProto.Entry = append(tagProto.Entry[:i], tagProto.Entry[i+1:]...)
			}
		}
	}
	evt.dirtyTags = false
}
